### PR TITLE
Search dropdown added in time-tracking

### PR DESCRIPTION
**Pull request by abinashpa**
## Notion card
https://saeloun.notion.site/Search-dropdown-added-in-time-tracking-684e48a96eb244c8a7af1a4315aacaf3

## Summary
Replaced the select elements with react-select.

## Preview
https://www.loom.com/share/e23d7c77d9f64d6fa1b4ec7749345bba

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [ ] This change requires a documentation update

### Checklist:

- [x] I have manually tested all workflows
- [x] I have performed a self-review of my own code
- [ ] I have added automated tests for my code

### CODE REVIEW

**Review by ajinkyaa — APPROVED** [review]
I see lot of duplication of code, can you please create common component out of this?

File: app/javascript/src/components/time-tracking/SelectProject.tsx [hunk 1]
```diff
@@ -17,6 +18,9 @@ const SelectProject: React.FC<Iprops> = ({
   isWeeklyEditing, // eslint-disable-line
   setIsWeeklyEditing
 }) => {
+  const clientList = clients.map(client => ({ value: client.name, label: client.name }));
```
Comment (onkar-saeloun) [inline app/javascript/src/components/time-tracking/SelectProject.tsx:21]: These maps will be looped each time you update the component. Better to keep it in useEffect along with specific conditions or wrapping it by a function..

**Review by onkar-saeloun — CHANGES_REQUESTED** [review]

**Comment by supriya3105** [comment]
@abinashpa Why is the title suggesting that search is added to time tracking. The Notion card linked to this PR is for search on projects list page.

File: app/javascript/src/common/SyncAutoComplete.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+  return (
+    <div className={`${size}-auto-complete-container`}>
+      <Autocomplete
+        getItemValue={(item) => item["value"]}
+        items={dropdownItems}
+        menuStyle={cssStyles.menuStyles}
+        value={searchValue}
+        onChange={(event) => {
+          setValue(event.target.value);
```
Comment (onkar-saeloun) [inline app/javascript/src/common/SyncAutoComplete.tsx:70]: please declare a function and add your logic over there just initiate the function here.
for example,
const handleOnChange = (e) => {
setValue(e.target.value)
}
<Autocomplete
  onChange={handleOnChange}
  ...
 />

File: app/javascript/src/common/SyncAutoComplete.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        }}
+        onSelect={(value) => {
```
Comment (onkar-saeloun) [inline app/javascript/src/common/SyncAutoComplete.tsx:72]: same as above

File: app/javascript/src/common/SyncAutoComplete.tsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            {item.label}
+          </div>
+        }
+        renderMenu={
+          function (items, value, style) {
+            if (items.length > 0) {
+              return <div style={{ ...style, ...this.menuStyle }} children={items} />; //eslint-disable-line
+            }
+            if (items.length === 0 && searchValue === "") {
+              return <></>;
```
Comment (onkar-saeloun) [inline app/javascript/src/common/SyncAutoComplete.tsx:87]: why do we need fragment over here `<></>`? Actually do we need this if condition?

File: app/javascript/src/components/time-tracking/AddEntry.tsx [hunk 5]
```diff
@@ -126,49 +131,28 @@ const AddEntry: React.FC<Iprops> = ({
...
-              </option>
-            )}
-            {client &&
-            projects[client].map((project, i) => (
-              <option data-project-id={project.id} key={i.toString()}>
-                {project.name}
-              </option>
-            ))}
-          </select>
+          {isDataFilled &&<SyncAutoComplete
```
Comment (onkar-saeloun) [inline app/javascript/src/components/time-tracking/AddEntry.tsx:149]: please format over here

File: app/controllers/time_tracking_controller.rb [hunk 6]
```diff
@@ -8,6 +8,7 @@ def index
     is_admin = current_user.has_role?(:owner, current_company) || current_user.has_role?(:admin, current_company)
     user_id = current_user.id
     employees = is_admin ? current_company.users.select(:id, :first_name, :last_name) : [current_user]
+    full_name = current_user.first_name + " " + current_user.last_name
```
Comment (onkar-saeloun) [inline app/controllers/time_tracking_controller.rb:11]: You can use template literals over here,
full_name = `${current_user.first_name} ${current_user.last_name}`

**Review by onkar-saeloun — CHANGES_REQUESTED** [review]
@abinashpa few small comments please have a look

File: app/javascript/src/common/SyncAutoComplete.tsx [hunk 7]
```diff
@@ -0,0 +1,111 @@
+import * as React from "react";
```
Comment (ajinkyaa) [inline app/javascript/src/common/SyncAutoComplete.tsx:1]: no need to import *, use `import React from "react"`

Comment (ajinkyaa) [inline app/javascript/src/common/SyncAutoComplete.tsx:1]: Add comment at the first line, that this file is for synchronous auto complete.

File: app/javascript/src/components/Invoices/MultipleEntriesModal/index.tsx [hunk 8]
```diff
@@ -91,7 +91,7 @@ const MultipleEntriesModal = ({
 
   return (
     <div style={{ background: "rgba(29, 26, 49,0.6)" }} className="px-52 py-20 w-full h-full fixed inset-0 flex justify-center z-50">
-      <div className="bg-white rounded-lg w-full h-128 flex flex-col justify-between">
+      <div className="bg-white rounded-lg w-full h-160 flex flex-col justify-between">
```
Comment (ajinkyaa) [inline app/javascript/src/components/Invoices/MultipleEntriesModal/index.tsx:94]: why do we need height? can we just keep the height as per the content?
what if we remove height from the modal?

File: app/javascript/src/common/SyncAutoComplete.tsx (hunk 4 above)
Comment (abinashpa) [reply app/javascript/src/common/SyncAutoComplete.tsx:87]: This code is inherited from AutoComplete.tsx.

**Review by onkar-saeloun — APPROVED** [review]

**Review by ajinkyaa — APPROVED** [review]

**Comment by apoorv1316** [comment]
@abinashpa Please add a notion card and the loom video.

**Comment by abinashpa** [comment]
> @abinashpa Please add a notion card and the loom video.

Notion card and Loom video has been added @apoorv1316.

**Review by apoorv1316 — CHANGES_REQUESTED** [review]
@abinashpa I cannot see projects just by clicking on the project field. Refer to loom: https://www.loom.com/share/1eb059b564e84c328262e6b3e7b6fe6d
 Also, can we add a placeholder text for the project & client text field?

**Comment by apoorv1316** [comment]
@abinashpa Can we show the current logged-in admin/owner's name when the time tracking page loads for the first time? The empty space doesn't look so good. Also, the workspace switcher is showing under the search bar. Can you fix it?
Screenshot 2022-07-06 at 5 22 23 PM
Screenshot 2022-07-06 at 5 24 35 PM

**Comment by abinashpa** [comment]
@apoorv1316 Fixes have been added

**Comment by apoorv1316** [comment]
@supriya3105 @abinashpa LGTM from functional perspective. Just wanted to confirm that this search box and results are as per the designs. Coz the search box in the middle looks a bit odd to me.
Screenshot 2022-07-08 at 1 18 51 PM

**Review by apoorv1316 — APPROVED** [review]

**Comment by abinashpa** [comment]
> @supriya3105 @abinashpa LGTM from functional perspective. Just wanted to confirm that this search box and results are as per the designs. Coz the search box in the middle looks a bit odd to me.
> Screenshot 2022-07-08 at 1 18 51 PM
> 

On figma designs search box has been moved to center.

**Comment by supriya3105** [comment]
@abinashpa  @apoorv1316 As per the designs it has been moved to centre. However we will need to make some changes to the search dropdown . Will add that as a separate ticket.  
Merging this PR for now.

**Pull request merged**